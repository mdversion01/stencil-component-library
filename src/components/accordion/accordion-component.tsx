import { Component, Prop, State, h, Event, EventEmitter, Watch } from '@stencil/core';

@Component({
  tag: 'accordion-component',
  styleUrl: 'accordion.scss',
  shadow: false,
})
export class AccordionComponent {
  @Prop() accordion = false;
  @Prop() contentTxtSize = '';
  @Prop() targetId = '';
  @Prop() classNames = '';
  @Prop({ reflect: true }) flush: boolean = false;
  @Prop() outlined = false;
  @Prop() block = false;
  @Prop() variant = '';
  @Prop() size = '';
  @Prop() disabled = false;
  @Prop() ripple = false;
  @Prop() link = false;
  @Prop() icon: string = 'fas fa-angle-down';
  @Prop() isOpen: boolean = false;

  @State() internalOpen = false;

  @Event() toggleEvent: EventEmitter<boolean>;

  @Watch('isOpen')
  watchIsOpen(newVal: boolean) {
    this.internalOpen = newVal;
    this.setInitialAccordionState();
  }

  componentWillLoad() {
    this.internalOpen = this.isOpen;
  }

  componentDidLoad() {
    this.setInitialAccordionState();
  }

  private setInitialAccordionState() {
    const targetEl = document.querySelector(`#${this.targetId}`) as HTMLElement;
    if (targetEl) {
      if (this.internalOpen) {
        targetEl.style.display = 'block';
        targetEl.style.height = 'auto';
      } else {
        targetEl.style.display = 'none';
        targetEl.style.height = '0px';
      }
    }
  }

  private getComputedClassAttribute() {
    let cls = this.classNames ? `${this.classNames}` : '';
    if (this.accordion) cls += ' accordion-button';
    if (this.ripple) cls += ' btn-ripple';
    if (this.size) cls += ` ${this.size}`;
    return cls.trim();
  }

  private textSizing() {
    switch (this.contentTxtSize) {
      case 'lg':
        return 'text-large';
      case 'default':
        return 'text-default';
      case 'sm':
        return 'text-small';
      case 'xs':
        return 'text-xsmall';
      case 'xl':
        return 'text-xlarge';
      case 'xxl':
        return 'text-xxlarge';
      default:
        return '';
    }
  }

  private toggleCollapse() {
    const targetEl = document.querySelector(`#${this.targetId}`) as HTMLElement;
    if (!targetEl) return;

    const isOpening = !this.internalOpen;
    this.internalOpen = isOpening;
    this.toggleEvent.emit(this.internalOpen);

    if (isOpening) {
      targetEl.style.display = 'block';
      targetEl.style.height = '0px';
      requestAnimationFrame(() => {
        targetEl.style.height = `${targetEl.scrollHeight}px`;
      });
    } else {
      targetEl.style.height = `${targetEl.scrollHeight}px`;
      requestAnimationFrame(() => {
        targetEl.style.height = '0px';
      });
    }

    const onTransitionEnd = () => {
      if (this.internalOpen) {
        targetEl.style.height = 'auto';
      } else {
        targetEl.style.display = 'none';
      }
      targetEl.removeEventListener('transitionend', onTransitionEnd);
    };
    targetEl.addEventListener('transitionend', onTransitionEnd);
  }

  get iconArray(): string[] {
    return this.icon ? this.icon.split(',').map(i => i.trim()) : [];
  }

  private handleKeyDown(event: KeyboardEvent) {
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();
      this.toggleCollapse();
    }
  }

  private renderExpansionContentArea() {
    return (
      <div
        id={this.targetId}
        role="region"
        aria-labelledby={`${this.targetId}-header`}
        class="collapse"
        style={{
          height: '0px',
          overflow: 'hidden',
          transition: 'height 0.35s ease',
        }}
      >
        <div class={`${this.accordion ? 'accordion-body' : 'accordion-card accordion-body'} ${this.textSizing()}`}>
          <slot name="content"></slot>
        </div>
      </div>
    );
  }

  private renderAccordionButton() {
    const displayIcon = this.iconArray.length === 1 ? this.iconArray[0] : this.internalOpen ? this.iconArray[1] : this.iconArray[0];

    return (
      <button-component
        onCustomClick={() => this.toggleCollapse()}
        aria-expanded={this.internalOpen ? 'true' : 'false'}
        aria-controls={this.targetId}
        classNames={`single ${this.getComputedClassAttribute()} ${!this.internalOpen ? 'collapsed' : ''}`}
        data-toggle="collapse"
        data-target={this.targetId}
        variant={this.variant}
        size={this.size}
        disabled={this.disabled}
        block={this.block}
      >
        <span class="accordion-label">
          <slot name="accordion-header"></slot>
          <span class={this.internalOpen ? 'rotate-down' : 'rotate-up'}>
            <icon-component icon={displayIcon}></icon-component>
          </span>
        </span>
      </button-component>
    );
  }

  private renderAccordionExpansionCard() {
    return (
      <div class={`accordion ${this.flush ? 'accordion-flush' : ''}`} role="presentation">
        <div class="accordion-item">
          {this.renderAccordionButton()}
          {this.renderExpansionContentArea()}
        </div>
      </div>
    );
  }

  private renderToggleButton() {
    return (
      <button-component
        onCustomClick={() => this.toggleCollapse()}
        onKeyDown={(e: KeyboardEvent) => this.handleKeyDown(e)}
        aria-expanded={this.internalOpen ? 'true' : 'false'}
        aria-controls={this.targetId}
        role="button"
        tabindex="0"
        classNames={`${this.getComputedClassAttribute()} ${!this.internalOpen ? 'collapsed' : ''}`}
        data-toggle="collapse"
        data-target={this.targetId}
        variant={this.variant}
        size={this.size}
        disabled={this.disabled}
        outlined={this.outlined}
        block={this.block}
      >
        <slot name="button-text"></slot>
      </button-component>
    );
  }

  private renderToggleLink() {
    return (
      <button-component
        link
        aria-expanded={this.internalOpen ? 'true' : 'false'}
        aria-controls={this.targetId}
        data-toggle="collapse"
        data-target={this.targetId}
        onCustomClick={() => this.toggleCollapse()}
        onKeyDown={(e: KeyboardEvent) => this.handleKeyDown(e)}
        url={`#${this.targetId}`}
        role="button"
        tabindex="0"
        variant={this.variant}
      >
        <slot name="button-text"></slot>
      </button-component>
    );
  }

  render() {
    if (this.accordion) {
      return this.renderAccordionExpansionCard();
    } else if (this.link) {
      return (
        <div class="accordion-wrapper">
          {this.renderToggleLink()}
          {this.renderExpansionContentArea()}
        </div>
      );
    } else {
      return (
        <div class="accordion-wrapper">
          {this.renderToggleButton()}
          {this.renderExpansionContentArea()}
        </div>
      );
    }
  }
}
