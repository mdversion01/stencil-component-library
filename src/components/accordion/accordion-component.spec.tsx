// src/components/accordion/accordion-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { AccordionComponent } from './accordion-component';

describe('accordion-component', () => {
  const getTargetEl = (page: any, id: string) =>
    page.root.querySelector<HTMLElement>(`#${id}`);

  const fireTransitionEnd = async (el: HTMLElement, page: any) => {
    el.dispatchEvent(new Event('transitionend'));
    await page.waitForChanges();
  };

  it('renders closed by default', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="section1"></accordion-component>`,
    });

    const button = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'section1');

    expect(button).toBeTruthy();
    expect(button.getAttribute('aria-expanded')).toBe('false');
    expect(target.style.display).toBe('none');
    expect(target.style.height).toBe('0px');

    expect(page.root).toMatchSnapshot();
  });

  it('renders open by default when is-open is set', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="section2" is-open="true"></accordion-component>`,
    });

    const button = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'section2');

    expect(page.rootInstance.internalOpen).toBe(true);
    expect(button.getAttribute('aria-expanded')).toBe('true');
    expect(target.style.display).toBe('block');
    expect(target.style.height).toBe('auto');

    expect(page.root).toMatchSnapshot();
  });

  it('toggles open/closed state on click', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="section3"></accordion-component>`,
    });

    const instance = page.rootInstance as AccordionComponent;
    const button = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'section3');

    await button.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(true);
    expect(button.getAttribute('aria-expanded')).toBe('true');
    await fireTransitionEnd(target, page);
    expect(target.style.height).toBe('auto');
    expect(target.style.display).toBe('block');

    await button.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(false);
    expect(button.getAttribute('aria-expanded')).toBe('false');
    await fireTransitionEnd(target, page);
    expect(target.style.display).toBe('none');
  });

  it('renders icon and switches when toggled', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component
               accordion
               target-id="accordion-test"
               icon="fa-chevron-down, fa-chevron-up"
             >
               <span slot="accordion-header">Header</span>
               <div slot="content">Body</div>
             </accordion-component>`,
    });

    const icon = page.root.querySelector('icon-component');
    expect(icon).toBeTruthy();
    expect(icon.getAttribute('icon')).toContain('fa-chevron-down');

    const button = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'accordion-test');

    button.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    await fireTransitionEnd(target, page);

    const updatedIcon = page.root.querySelector('icon-component');
    expect(updatedIcon.getAttribute('icon')).toContain('fa-chevron-up');
  });

  it('toggles open/closed with keyboard (Enter and Space)', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="section5"></accordion-component>`,
    });

    const instance = page.rootInstance as AccordionComponent;
    const button = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'section5');

    await button.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(true);
    await fireTransitionEnd(target, page);

    await button.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(false);
    await fireTransitionEnd(target, page);
  });

  it('renders and toggles in link mode (anchor-like button)', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component link target-id="link-section" variant="primary">
               <span slot="button-text">Open Link</span>
             </accordion-component>`,
    });

    const instance = page.rootInstance as AccordionComponent;
    const linkBtn = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'link-section');

    expect(linkBtn.getAttribute('role')).toBe('button');
    expect(linkBtn.getAttribute('aria-expanded')).toBe('false');

    linkBtn.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(true);
    expect(linkBtn.getAttribute('aria-expanded')).toBe('true');
    await fireTransitionEnd(target, page);

    await linkBtn.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(false);
    await fireTransitionEnd(target, page);
  });

  it('emits toggleEvent(boolean) with the new state', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="emit-section"></accordion-component>`,
    });

    const spy = jest.fn();
    page.root.addEventListener('toggleEvent', (e: CustomEvent) => spy(e.detail));
    const btn = page.root.querySelector('button-component');

    await btn.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    expect(spy).toHaveBeenLastCalledWith(true);

    await btn.dispatchEvent(new CustomEvent('customClick'));
    await page.waitForChanges();
    expect(spy).toHaveBeenLastCalledWith(false);
  });

  it('@Watch("isOpen"): updating host prop updates internalOpen and styles (no prop mutation on instance)', async () => {
    const page = await newSpecPage({
      components: [AccordionComponent],
      html: `<accordion-component target-id="watched"></accordion-component>`,
    });

    const instance = page.rootInstance as AccordionComponent;
    const btn = page.root.querySelector('button-component');
    const target = getTargetEl(page, 'watched');

    // Start closed
    expect(instance.internalOpen).toBe(false);
    expect(btn.getAttribute('aria-expanded')).toBe('false');

    // Open via host attribute/property (external change)
    page.root.setAttribute('is-open', 'true'); // OR: (page.root as any).isOpen = true;
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(true);
    expect(btn.getAttribute('aria-expanded')).toBe('true');
    expect(target.style.display).toBe('block');
    expect(target.style.height).toBe('auto');

    // Close via host attribute/property
    page.root.setAttribute('is-open', 'false'); // OR: (page.root as any).isOpen = false;
    await page.waitForChanges();
    expect(instance.internalOpen).toBe(false);
    expect(btn.getAttribute('aria-expanded')).toBe('false');
    expect(target.style.display).toBe('none');
    expect(target.style.height).toBe('0px');
  });
});
