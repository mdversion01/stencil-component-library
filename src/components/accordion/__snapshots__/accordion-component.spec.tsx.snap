// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accordion-component renders closed by default 1`] = `
<accordion-component target-id="section1">
  <!---->
  <div class="accordion-wrapper">
    <button-component aria-controls="section1" aria-expanded="false" classnames=" collapsed" data-target="section1" data-toggle="collapse" role="button" size="" tabindex="0" variant=""></button-component>
    <div aria-labelledby="section1-header" class="collapse" id="section1" role="region" style="height: 0px; overflow: hidden; transition: height 0.35s ease; display: none;">
      <div class="accordion-body accordion-card"></div>
    </div>
  </div>
</accordion-component>
`;

exports[`accordion-component renders open by default when is-open is set 1`] = `
<accordion-component is-open="true" target-id="section2">
  <!---->
  <div class="accordion-wrapper">
    <button-component aria-controls="section2" aria-expanded="true" classnames=" " data-target="section2" data-toggle="collapse" role="button" size="" tabindex="0" variant=""></button-component>
    <div aria-labelledby="section2-header" class="collapse" id="section2" role="region" style="height: auto; overflow: hidden; transition: height 0.35s ease; display: block;">
      <div class="accordion-body accordion-card"></div>
    </div>
  </div>
</accordion-component>
`;
